### PR TITLE
Add workspace background color option

### PR DESCRIPTION
**Pull request by soreau**
Fixes wf-shell #26.

### CODE REVIEW

**Comment by soreau** [comment]
> A few issues:
> 
>     1. Why does this all originate in output-layout? Can't it be contained inside render-manager?
I figured all the code would be called once per output if was done in render_manager, eg. get_option and setting up the changed callback.

> 
>     2. This won't work for custom workspace streams, for ex. the ones that expo uses.
I see you are correct, though I'm not sure why.

> 
> 
> So what I suggest: Set the default workspace color (in workspace-stream.hpp) to an invalid color. When clearing the background, check whether the color is invalid, and then use the default color (so that plugins can still override background color if they wish so).

I don't really understand what you mean here.

**Comment by ammen99** [comment]
> > A few issues:
> > ```
> > 1. Why does this all originate in output-layout? Can't it be contained inside render-manager?
> > ```
> 
> I figured all the code would be called once per output if was done in render_manager, eg. get_option and setting up the changed callback.
> 

It's okay to make this multiple times, changing a config option isn't a costly operation (and doesn't happen every frame)

> > ```
> > 2. This won't work for custom workspace streams, for ex. the ones that expo uses.
> > ```
> 
> I see you are correct, though I'm not sure why.
Because the plugin can create its own workspace stream. And you set the background for only default workspace streams.

> 
> > So what I suggest: Set the default workspace color (in workspace-stream.hpp) to an invalid color. When clearing the background, check whether the color is invalid, and then use the default color (so that plugins can still override background color if they wish so).
> 
> I don't really understand what you mean here.

Maybe the comment about having non-default workspaces would be helpful in understanding. 
Let me try to explain it in another way: plugins can create custom workspaces and can request custom background. We just need some way for plugins to signal when they want the default background color and when they want to override it. So I proposed, setting the `stream.background` by default to some invalid value, then have render-manager take the default color in this case when updating the workspace stream.

**Comment by soreau** [comment]
I think I see what you mean now. I have it working, let me know if it looks more inline now.

File: src/api/workspace-stream.hpp [hunk 1]
```diff
@@ -18,7 +18,7 @@ struct workspace_stream_t
     float scale_y = 1.0;
 
     /* The background color of the stream, when there is no view above it */
```
Comment (ammen99) [inline src/api/workspace-stream.hpp]: Maybe explain what the negative alpha represents (in the comment)

File: src/output/render-manager.cpp [hunk 2]
```diff
@@ -864,7 +888,12 @@ class wf::render_manager::impl
         }
 
         check_schedule_surfaces(repaint, stream);
-        clear_empty_areas(repaint, stream.background);
```
Comment (ammen99) [inline src/output/render-manager.cpp]: braces around this

File: src/output/render-manager.cpp [hunk 3]
```diff
@@ -864,7 +888,12 @@ class wf::render_manager::impl
...
+
+        if (stream.background.a < 0)
```
Comment (ammen99) [inline src/output/render-manager.cpp]: Braces around if/else

File: src/output/render-manager.cpp [hunk 4]
```diff
@@ -348,6 +353,18 @@ class wf::render_manager::impl
...
+
+        background_color_opt_changed = [=] ()
+        {
+            auto color = background_color_opt->as_color();
+            update_background_color(color);
+        };
+
+        auto section = wf::get_core().config->get_section("core");
+        background_color_opt = section->get_option("background_color", "0 0 0 1");
+        background_color_opt->add_updated_handler(&background_color_opt_changed);
```
Comment (ammen99) [inline src/output/render-manager.cpp]: One more thing: when the render manager is destroyed, we need to disconnect the updated handler.

**Comment by ammen99** [comment]
This PR LGTM. Thanks!

**Pull request merged**